src/audio/module_adapter/module_adapter.c

```diff
@@ -623,54 +630,126 @@ static void module_adapter_process_output(struct comp_dev *dev)
 	}
 }
 
+static uint32_t
+module_single_sink_setup(struct comp_dev *dev,
+			 struct comp_buffer __sparse_cache **source_c,
+			 struct comp_buffer __sparse_cache **sinks_c)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct comp_copy_limits c;
+	struct list_item *blist;
+	uint32_t num_input_buffers = 0;
+	int i = 0;
+
+	list_for_item(blist, &dev->bsource_list) {
+		/* check if the source dev is in the same state as the dev */
+		if (!source_c[i]->source || source_c[i]->source->state != dev->state) {
```

Comment:
I still don't like this. We're in `.copy()` i.e. in the pipeline task, actually performing streaming, right? These buffers can only be modified when modifying topology, i.e. either when connecting or when disconnecting pipelines. Those actions should also only be done in a pipeline task context. As long as we only allow connected pipelines on the same core, these tasks will run sequentially without preempting each other. So within one scheduling invocation these buffers remain consistent.
Only when we move to connecting different cores, then this will become a problem. But even then this will be a larger problem, requiring proper robust locking. And if we ever hit this condition, it means we were in an inconsistent state and it can bite us elsewhere. These checks seem only to reduce the probability of hitting a problem, making it more difficult to debug and fix. Really don't seem useful to me. Same below of course.

Reply:
@lyakh I'm fully with you that this does not seem right in copy (we should have less frequent state change callbacks where module can react to changes in its configuration or the configuration of its immediate connections, and NOT do this check in each copy), but it's still a fact that currently this is hit in pause-resume tests (with sof-tests), so it's not like this is a corner case at this point.

Reply:
oh i missed these 2. fixed now

Reply:
> @lyakh I'm fully with you that this does not seem right in copy (we should have less frequent state change callbacks where module can react to changes in its configuration or the configuration of its immediate connections, and NOT do this check in each copy), but it's still a fact that currently this is hit in pause-resume tests (with sof-tests), so it's not like this is a corner case at this point.

@kv2019i if it's so easy to hit, then we really need to investigate and fix these properly. Maybe someone could add a task to our planning ;-)